File: src/adapters/kicad/libraries.ts
import { existsSync, readdirSync, readFileSync } from "node:fs";
import { homedir } from "node:os";
import { dirname, extname, join } from "node:path";

import { args, child, children, parseSexpr } from "./sexpr.js";

export type LibraryEntry = { name: string; uri: string; type: string };

/** KiCad install roots, newest first. */
function kicadRoots(): string[] {
  const roots: string[] = [];
  const base = "C:/Program Files/KiCad";
  if (existsSync(base)) {
    for (const version of readdirSync(base).sort((a, b) =>
      b.localeCompare(a, "en", { numeric: true }),
    )) {
      roots.push(join(base, version));
    }
  }
  for (const unix of ["/usr/share/kicad", "/usr/local/share/kicad"]) {
    if (existsSync(unix)) roots.push(dirname(dirname(unix)));
  }
  return roots;
}

/**
 * Resolves `${KICAD9_FOOTPRINT_DIR}` and friends.
 *
 * These rarely exist in the process environment — KiCad keeps them in
 * `kicad_common.json` and falls back to built-in defaults — and a KiCad 10
 * install can still carry a table written by KiCad 9, so the version in the
 * variable name is deliberately ignored when falling back.
 */
export function resolveVars(uri: string): string {
  return uri.replace(/\$\{([^}]+)\}/g, (whole, name: string) => {
    const fromEnv = process.env[name];
    if (fromEnv) return fromEnv;

    const fromConfig = configVars()[name];
    if (fromConfig) return fromConfig;

    const suffix = /FOOTPRINT_DIR$/.test(name)
      ? "share/kicad/footprints"
      : /SYMBOL_DIR$/.test(name)
        ? "share/kicad/symbols"
        : /3DMODEL_DIR$/.test(name)
          ? "share/kicad/3dmodels"
          : undefined;
    if (suffix) {
      for (const root of kicadRoots()) {
        const candidate = join(root, suffix);
        if (existsSync(candidate)) return candidate;
      }
    }
    return whole;
  });
}

let configCache: Record<string, string> | undefined;

function configVars(): Record<string, string> {
  if (configCache) return configCache;
  configCache = {};
  for (const dir of kicadConfigDirs()) {
    const file = join(dir, "kicad_common.json");
    if (!existsSync(file)) continue;
    try {
      const parsed: unknown = JSON.parse(readFileSync(file, "utf8"));
      const vars =
        typeof parsed === "object" && parsed !== null
          ? (parsed as { environment?: { vars?: Record<string, string> } }).environment?.vars
          : undefined;
      if (vars) configCache = { ...vars, ...configCache };
    } catch {
      // a malformed config is not this tool's problem to report
    }
  }
  return configCache;
}

/** KiCad's per-user config directories, newest version first. */
export function kicadConfigDirs(): string[] {
  const roots =
    process.platform === "win32"
      ? [join(process.env["APPDATA"] ?? join(homedir(), "AppData/Roaming"), "kicad")]
      : process.platform === "darwin"
        ? [join(homedir(), "Library/Preferences/kicad")]
        : [join(process.env["XDG_CONFIG_HOME"] ?? join(homedir(), ".config"), "kicad")];

  const dirs: string[] = [];
  for (const root of roots) {
    if (!existsSync(root)) continue;
    for (const version of readdirSync(root).sort((a, b) =>
      b.localeCompare(a, "en", { numeric: true }),
    )) {
      const dir = join(root, version);
      if (existsSync(join(dir, "fp-lib-table"))) dirs.push(dir);
    }
  }
  return dirs;
}

export function parseLibTable(text: string): LibraryEntry[] {
  const root = parseSexpr(text);
  return children(root, "lib").map((lib) => ({
    name: args(child(lib, "name"))[0] ?? "",
    uri: resolveVars(args(child(lib, "uri"))[0] ?? ""),
    type: args(child(lib, "type"))[0] ?? "KiCad",
  }));
}

/**
 * Footprint libraries visible to a project: the project's own table first, so
 * a project-local library shadows a global one of the same nickname, exactly
 * as KiCad resolves them.
 */
export function libraryTable(projectPath: string): Map<string, LibraryEntry> {
  const table = new Map<string, LibraryEntry>();

  for (const dir of kicadConfigDirs()) {
    for (const entry of parseLibTable(readFileSync(join(dir, "fp-lib-table"), "utf8"))) {
      if (!table.has(entry.name)) table.set(entry.name, entry);
    }
  }

  const projectDir = extname(projectPath) === "" ? projectPath : dirname(projectPath);
  const projectTable = join(projectDir, "fp-lib-table");
  if (existsSync(projectTable)) {
    for (const entry of parseLibTable(readFileSync(projectTable, "utf8"))) {
      table.set(entry.name, entry); // project wins
    }
  }

  return table;
}

/** Absolute path of `Library:Footprint`, or undefined when it cannot be found. */
export function resolveFootprint(
  reference: string,
  table: Map<string, LibraryEntry>,
): string | undefined {
  const colon = reference.indexOf(":");
  if (colon < 0) return undefined;

  const library = table.get(reference.slice(0, colon));
  if (!library) return undefined;

  const path = join(library.uri, `${reference.slice(colon + 1)}.kicad_mod`);
  return existsSync(path) ? path : undefined;
}

File: kinv/adapters/kicad/libraries.py
"""Footprint library tables: where ``Library:Footprint`` lives on disk."""

from __future__ import annotations

import json
import os
import re
import sys
from pathlib import Path

from kinv.adapters.kicad.hierarchy import read_text
from kinv.adapters.kicad.sexpr import args, child, children, parse_sexpr
from kinv.js import collation_key


def _newest_first(names: list[str]) -> list[str]:
    return sorted(names, key=lambda v: collation_key(v, numeric=True), reverse=True)


def _kicad_roots() -> list[str]:
    """KiCad install roots, newest first."""
    roots: list[str] = []
    base = "C:/Program Files/KiCad"
    if os.path.exists(base):
        roots.extend(os.path.join(base, version) for version in _newest_first(os.listdir(base)))
    for unix in ("/usr/share/kicad", "/usr/local/share/kicad"):
        if os.path.exists(unix):
            roots.append(os.path.dirname(os.path.dirname(unix)))
    return roots


_VAR = re.compile(r"\$\{([^}]+)\}")


def resolve_vars(uri: str) -> str:
    """Resolves ``${KICAD9_FOOTPRINT_DIR}`` and friends.

    These rarely exist in the process environment — KiCad keeps them in
    ``kicad_common.json`` and falls back to built-in defaults — and a KiCad 10
    install can carry a table written by KiCad 9, so the version in the
    variable name is deliberately ignored when falling back.
    """

    def replace(match: re.Match[str]) -> str:
        name = match.group(1)
        from_env = os.environ.get(name)
        if from_env:
            return from_env
        from_config = _config_vars().get(name)
        if from_config:
            return from_config

        suffix = (
            "share/kicad/footprints"
            if name.endswith("FOOTPRINT_DIR")
            else "share/kicad/symbols"
            if name.endswith("SYMBOL_DIR")
            else "share/kicad/3dmodels"
            if name.endswith("3DMODEL_DIR")
            else None
        )
        if suffix:
            for root in _kicad_roots():
                candidate = os.path.join(root, suffix)
                if os.path.exists(candidate):
                    return candidate
        return match.group(0)

    return _VAR.sub(replace, uri)


_config_cache: dict[str, str] | None = None


def _config_vars() -> dict[str, str]:
    global _config_cache
    if _config_cache is not None:
        return _config_cache
    merged: dict[str, str] = {}
    for directory in kicad_config_dirs():
        file = os.path.join(directory, "kicad_common.json")
        if not os.path.exists(file):
            continue
        try:
            parsed = json.loads(read_text(file))
            variables = parsed.get("environment", {}).get("vars") if isinstance(parsed, dict) else None
            if isinstance(variables, dict):
                # the newest version's config, read first, wins
                merged = {**variables, **merged}
        except (OSError, ValueError, AttributeError):
            pass  # a malformed config is not this tool's problem to report
    _config_cache = merged
    return merged


def kicad_config_dirs() -> list[str]:
    """KiCad's per-user config directories, newest version first."""
    home = str(Path.home())
    if sys.platform == "win32":
        roots = [os.path.join(os.environ.get("APPDATA") or os.path.join(home, "AppData/Roaming"), "kicad")]
    elif sys.platform == "darwin":
        roots = [os.path.join(home, "Library/Preferences/kicad")]
    else:
        roots = [os.path.join(os.environ.get("XDG_CONFIG_HOME") or os.path.join(home, ".config"), "kicad")]

    dirs: list[str] = []
    for root in roots:
        if not os.path.exists(root):
            continue
        for version in _newest_first(os.listdir(root)):
            directory = os.path.join(root, version)
            if os.path.exists(os.path.join(directory, "fp-lib-table")):
                dirs.append(directory)
    return dirs


def parse_lib_table(text: str) -> list[dict]:
    root = parse_sexpr(text)
    return [
        {
            "name": (args(child(lib, "name")) or [""])[0],
            "uri": resolve_vars((args(child(lib, "uri")) or [""])[0]),
            "type": (args(child(lib, "type")) or ["KiCad"])[0],
        }
        for lib in children(root, "lib")
    ]


def library_table(project_path: str) -> dict[str, dict]:
    """Footprint libraries visible to a project; the project's own table shadows the global one."""
    table: dict[str, dict] = {}
    for directory in kicad_config_dirs():
        for entry in parse_lib_table(read_text(os.path.join(directory, "fp-lib-table"))):
            table.setdefault(entry["name"], entry)

    project_dir = project_path if os.path.splitext(project_path)[1] == "" else os.path.dirname(project_path)
    project_table = os.path.join(project_dir, "fp-lib-table")
    if os.path.exists(project_table):
        for entry in parse_lib_table(read_text(project_table)):
            table[entry["name"]] = entry  # project wins
    return table


def resolve_footprint(reference: str, table: dict[str, dict]) -> str | None:
    """Absolute path of ``Library:Footprint``, or None when it cannot be found."""
    colon = reference.find(":")
    if colon < 0:
        return None
    library = table.get(reference[:colon])
    if not library:
        return None
    path = os.path.join(library["uri"], f"{reference[colon + 1:]}.kicad_mod")
    return path if os.path.exists(path) else None
